src/Symfony/Component/Validator/Constraints/LengthValidator.php

```diff
@@ -40,23 +40,25 @@ public function validate($value, Constraint $constraint)
             $charset = 'UTF-8';
         }
 
-        if (function_exists('iconv_strlen')) {
-            $length = @iconv_strlen($stringValue, $constraint->charset);
-            $invalidCharset = false === $length;
+        if ('UTF-8' === $charset) {
+            if (!preg_match('//u', $stringValue)) {
+                $invalidCharset = true;
+            } elseif (function_exists('utf8_decode')) {
+                $length = strlen(utf8_decode($stringValue));
+            } else {
+                preg_replace('/./u', '', $stringValue, -1, $length);
+            }
         } elseif (function_exists('mb_strlen')) {
```

Comment:
moving mbstring before iconv because it is more reliable (iconv implementations varies between libiconv and glibc)